src/main/java/org/jpeek/metrics/cohesion/MMAC.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    /**
+     * Calculates MMAC metric for a single Java class.
+     *
+     * @param ctc The .class file
+     * @return MMAC metric
+     * @checkstyle ReturnCountCheck (10 lines). This assert is suppressed
+     *  because we have non one line formula for metric calculation.
+     * @checkstyle TrailingCommentCheck (50 lines)
+     * @checkstyle MethodBodyCommentsCheck (50 lines)
+     */
+    @SuppressWarnings(
+        // These PMD asserts are suppressed because we have non one line
+        // formula for metric calculation.
+        {
+            "PMD.OnlyOneReturn", "PMD.CyclomaticComplexity",
+            "PMD.StdCyclomaticComplexity", "PMD.ModifiedCyclomaticComplexity"
+        })
+    private static double cohesion(final CtClass ctc) {
+        final List<Collection<String>> methods = new ListOf<>(
+            new Mapped<>(
+                new Mapped<>(
+                    new Filtered<>(
+                        new IterableOf<>(ctc.getDeclaredMethods()),
+                        // isn't lambda or other generated methods by compiler
+                        m -> !m.getName().contains("$")
+                    ),
+                    CtBehavior::getSignature
+                ),
+                signature -> {
+                    final Set<String> types = new HashSet<>();
+                    new SignatureReader(signature).accept(
+                        new SignatureVisitor(Opcodes.ASM6) {
+                            @Override
+                            public void visitClassType(final String name) {
+                                super.visitClassType(name);
+                                types.add(name);
+                            }
+                            @Override
+                            public void visitBaseType(final char name) {
+                                super.visitBaseType(name);
+                                if ('V' != name) { // isn't void type
+                                    types.add(String.valueOf(name));
+                                }
+                            }
+                        }
+                    );
+                    return types;
+                }
+            )
+        );
+        if (methods.isEmpty()) {
+            return 0.0d;
+        }
+        if (methods.size() == 1) {
+            return 1.0d;
+        }
+        final Set<String> types = new HashSet<>();
+        for (final Collection<String> method : methods) {
+            types.addAll(method);
+        }
+        if (types.isEmpty()) {
+            return 0.0d;
+        }
+        double sum = 0.0d;
+        for (final String type : types) {
+            int mcnt = 0;
+            for (final Collection<String> method : methods) {
+                if (method.contains(type)) {
+                    ++mcnt;
+                }
+            }
+            sum += mcnt * (mcnt - 1);
+        }
+        return sum
+            / (types.size() * methods.size() * (methods.size() - 1));
```

Comment:
I dont think this is correct, it should be `1- (2/(methods.size() * (methods.size() - 1))*sum`

Reply:
Hmm, you partially right. Original formula for _MMAC_ is:
<img width="569" alt="mmac" src="https://user-images.githubusercontent.com/6314696/32343547-e6120a08-c014-11e7-9276-be93cb43a8fc.png">

See please [this one doc](http://waset.org/publications/5239/a-design-based-cohesion-metric-for-object-oriented-classes), chapter IV, paragraph "A. MMAC and AAC Metrics".

In your formula `sum` is sum of normalized distances between pair of methods, but I don't calculate normalized distances to `sum` variable. I calculate sum of methods which use the same types of parameters or return types and then divide this _sum_ to multiplication of distinct types and methods. It is another analogous way to calculate _MMAC_ metric. And it is described in mentioned above doc. See please formula (7) in doc
<img width="610" alt="mmac_7" src="https://user-images.githubusercontent.com/6314696/32343485-acfe99a2-c014-11e7-9f8a-51bb09126c08.png">
 and how we can get this one formula from original formula (original formula is presented as (2) in doc).

Reply:
Correct, thank you. 

@yegor256 You can merge this pr.